File: p23_ant/ant.cpp
#include <stdio.h>
#include <algorithm>

using namespace std;

int main(){
  int l = 0;
  int n = 0;
  int* ants = new int[1000000];

  scanf("%d", &l);
  scanf("%d", &n);
  for(int i=0; i < n; i++){
    scanf("%d", ants + i);
  }

  int minT = 0;
  for(int i=0; i < n; i++){
    minT = max(minT, min(ants[i], l - ants[i]));
  }

  int maxT = 0;
  for(int i=0; i < n; i++){
    maxT = max(maxT, max(ants[i], l - ants[i]));
  }

  printf("min = %d\n", minT);
  printf("max = %d\n", maxT);
  delete ants;

  return 0;
}

File: p23_ant/ant.hs
import Control.Monad.Reader
import Control.Monad.Writer
import Control.Applicative

data GivenData = GivenData{l :: Int, ants :: [Int]}

calc :: Reader GivenData (Int, Int)
calc = do
  l <- l <$> ask
  ants <- ants <$> ask
  let a = (min <$> (l - ) <*> id) <$> ants
  let b = (max <$> (l - ) <*> id) <$> ants
  return $ (maximum a, maximum b)

output :: (Int, Int) -> Writer String ()
output x = do
  tell $ "min = "
  tell $ show $ fst x
  tell $ "\n"
  tell $ "max = "
  tell $ show $ snd x
  return ()

main = do
  l <- (read :: String -> Int) <$> getLine
  _ <- getLine
  ants <- (((read :: String -> Int) <$>).words) <$> getLine
  let calced = runReader calc GivenData{l = l, ants = ants}
  putStrLn $ execWriter $ output calced
     